tests/integration/istio_test.go

```diff
@@ -0,0 +1,496 @@
+
+		istioVersion = "latest"
+		oldVersion   = "v1.20.0"
+		newVersion   = "v1.20.1" // TODO: get these versions from versions.yaml
+
+		gracePeriod = 30 * time.Second
+		pilotImage  = "maistra.io/test:latest"
+	)
+	istioKey := client.ObjectKey{Name: istioName}
+	istio := &v1alpha1.Istio{}
+
+	SetDefaultEventuallyTimeout(30 * time.Second)
+	SetDefaultEventuallyPollingInterval(time.Second)
+
+	SetDefaultConsistentlyDuration(30 * time.Second)
+	SetDefaultConsistentlyPollingInterval(time.Second)
+
+	ctx := context.Background()
+
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: istioNamespace,
+		},
+	}
+
+	BeforeAll(func() {
+		Step("Creating the Namespace to perform the tests")
+		Expect(k8sClient.Create(ctx, namespace)).To(Succeed())
+	})
+
+	AfterAll(func() {
+		// TODO(user): Attention if you improve this code by adding other context test you MUST
+		// be aware of the current delete namespace limitations.
+		// More info: https://book.kubebuilder.io/reference/envtest.html#testing-considerations
+		Step("Deleting the Namespace to perform the tests")
+		Expect(k8sClient.Delete(ctx, namespace)).To(Succeed())
+
+		deleteAllIstiosAndRevisions(ctx)
+	})
+
+	Describe("basic operation", func() {
+		BeforeAll(func() {
+			Step("Creating the custom resource")
+			istio = &v1alpha1.Istio{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: istioName,
+				},
+				Spec: v1alpha1.IstioSpec{
+					Version:   istioVersion,
+					Namespace: istioNamespace,
+					UpdateStrategy: &v1alpha1.IstioUpdateStrategy{
+						Type: v1alpha1.UpdateStrategyTypeInPlace,
+						InactiveRevisionDeletionGracePeriodSeconds: ptr.Of(int64(gracePeriod.Seconds())),
+					},
+					Values: []byte(`{
+						"pilot":{"image":"` + pilotImage + `"},
+						"istio_cni":{"enabled":true}
+					}`),
+				},
+			}
+			Expect(k8sClient.Create(ctx, istio)).To(Succeed())
+		})
+
+		It("updates the Istio resource", func() {
+			Eventually(func(g Gomega) {
+				g.Expect(k8sClient.Get(ctx, istioKey, istio)).To(Succeed())
+				g.Expect(istio.Status.ObservedGeneration).To(Equal(istio.Generation))
+			}).Should(Succeed())
+		})
+
+		It("creates the IstioRevision resource", func() {
+			revKey := client.ObjectKey{Name: istioName}
+			rev := &v1alpha1.IstioRevision{}
+			Eventually(func(g Gomega) {
+				g.Expect(k8sClient.Get(ctx, revKey, rev)).To(Succeed())
+				g.Expect(rev.GetOwnerReferences()).To(ContainElement(NewOwnerReference(istio)))
+			}).Should(Succeed())
+
+			Expect(rev.Spec).To(Equal(v1alpha1.IstioRevisionSpec{
+				Version:   istio.Spec.Version,
+				Namespace: istio.Spec.Namespace,
+				Values: deprettify([]byte(`{
+					"defaultRevision":"",
+					"gateways":{
+						"istio-egressgateway":{},
+						"istio-ingressgateway":{}
+					},
+					"global":{
+						"configValidation":true,
+						"istioNamespace":"` + istio.Spec.Namespace + `"
+					},
+					"istio_cni": {
+						"enabled":true
+					},
+					"pilot":{
+						"image":"maistra.io/test:latest"
```

Comment:
we should probably construct this string from the variables above- or at least `pilotImage` should be used, right?

Reply:
done